wsgi_kerberos.py

```diff
@@ -51,28 +58,29 @@ def _consume_request(environ, read_max):
         return
     if environ.get("HTTP_EXPECT") == "100-continue":
         return
+    if environ["REQUEST_METHOD"] in _NEVER_READ_METHODS:
+        return
     try:
-        sock = environ.get('wsgi.input')
-        if hasattr(sock, 'closed') and sock.closed:
+        body = environ.get('wsgi.input')
+        if hasattr(body, 'closed') and body.closed:
             return
 
         content_length = environ.get('CONTENT_LENGTH', '')
         if not content_length:
-            remaining = read_max
-        else:
-            content_length = int(content_length)
-            if content_length > read_max:
-                # User is not willing to read such a large request body, but
-                # reading anything less does not help naively-written clients.
-                return
-            remaining = content_length
+            return
+        content_length = int(content_length)
+        if content_length > read_max:
+            # User is not willing to read such a large request body, but
+            # reading anything less does not help naively-written clients.
+            return
```

Comment:
Should we raise a 413 response here? That's what other web servers do when the configured max body length is exceeded.

Reply:
We want to continue to send 401 even in this case. Sending 413 would cause a client whose request did not include credentials to not retry with credentials. (Clients automatically retry on 401, but not on 413.) Given that this _consume_request function is only a best-effort workaround for naive clients, I think it should try not to introduce any semantic changes.

Reply:
You're right - 413 is not correct here. However any client that doesn't send 100-Continue would likely get a broken pipe error (I've also tested golang http.Client, which intermittently exhibits this behavior). So maybe we should at least log an error here (suggesting to increase `read_max_on_auth_fail`)?

Reply:
Thanks, added more logging, and this also reminded me that I meant to update the Changelog to include additional guidance about these changes (and in particular how they affect requests_kerberos, which can set `force_preemptive=True` to avoid connection errors). Done in the latest revision. The new guidance also notes that users can set the wsgi_kerberos logger to debug level to view logs about requests from such clients.

Note: Rather than log this as an error, I think it's better to log at debug and tell users how to turn on debug logging (as in the latest revision), since this is not an error on the server side.

----

(Nice you were able to test with golang http.Client! Strange it exhibits this behavior intermittently, unless: by intermittently, do you mean it always exhibits this behavior with bodies larger than `read_max_on_auth_fail`, and never does with smaller bodies? I'd expect the behavior to be deterministic in each case.)

(Also, I'm sure not sending `Expect: 100-continue` is correlated with the naive behavior, but FWIW that's an independent optimization; sending `Expect: 100-continue` is not required for a client to avoid the naive behavior that results in broken pipe errors.)

Reply:
The silent failure mode when running into this issue is my biggest concern here. Your average developer/ops person is simply not gonna go around reading wsgi_kerberos changelog. What they will see is intermittent failures uploading large files, and have no idea where it's coming from. The failure needs to be loud and clear. I'm thinking something like this:
```
LOG.error("wsgi_kerberos discarded {content_size} bytes of request body. Consider increasing read_max_on_auth_failure or enable pre-emptive authentication on the client.")
```

Reply:
The golang client does not send a 100-Continue by default, but appears to fail intermittently for file size > max. I'm not really sure why it's failing intermittently (like 25% of the time), but when it does, it fails with a broken pipe error, which is what I'd expect. I'm using the example from https://stackoverflow.com/a/51455855.

Reply:
I’m sympathetic to that.

But on the flip side:

- Logging at "error" could also create noise, where the operator of the service suddenly starts seeing these error logs, and does not consider this an error, but rather desired behavior. And the only thing they can do to suppress the noisy logs is to crank our logger's level up past "error" to "critical", or disable our logger completely, both of which would result in missing the other messages we log at "error" level (e.g. "Unhandled GSSError"), which we really want them to see. (Server error logs should be used for actual server errors, not client errors.)
- The behavior that we changed occurred with a major version bump. In my experience, users actually are aware that upgrading to new major versions of dependencies can mean breaking changes, and realize they should read the changelog before upgrading, try to limit the impact of the upgrade (staged rollout, etc.), and otherwise take a little more care. (Not only that, but WSGI-Kerberos's previous behavior was released under a [major version zero](https://semver.org/#spec-item-4) version, which is specified as "Anything MAY change at any time. The public API SHOULD NOT be considered stable.")
- The changelog here really is not that much to read, is included right in the README, the package long_description, and in the GitHub Release notes  (thanks for adding those!) which get emailed to Watchers. And with the latest changes in this PR, we call particular attention to this potential issue.

Rather than changing the log level to error here, what about calling out the potential issue even more prominently in the changelog in a "potentially-breaking changes" sub-heading?

Reply:
You think too highly of the average user :). It doesn't matter how prominent the changelog is - it's highly unlikely that most people will read it or even notice the major version change. This library is a fairly minor component of most projects. For new projects, this is not even a change - just a corner case of which there is maybe a 1% chance that they'll ever run into. Error logging is fairly standard for this type of error. For example, in Nginx, if you send too large post body, it will log an error saying `client intended to send too large body: 53883021 bytes` - so in this case it is technically a client error, but is still logged as an error in the logs. If if turns out that users don't care about this error, they can also just disable read_max_on_auth_failure altogether. It's unlikely that this condition is being hit without causing client-side exceptions.

Reply:
Updated the message logged and made the level "error" in the latest revision.

I also added a time limit as suggested by @mkomitee in #27, with a default of 30 seconds. This seems especially worth having now that we're increasing the size limit to 200 MB. Included passing tests. What do you think?

Reply:
I could see this being useful *instead of* the size limit, but having both feels a bit too like scope creep for me. Nginx and apache both have sophisticated mechanisms to prevent this type of attack, and I feel like wsgi_kerberos is not the right place to implement this type of logic.

Reply:
The time-based limit is orthogonal to the size-based limit. Without the time-based limit, a client that sends e.g. one byte every 30 seconds will never trigger the size-based limit, but will cause the server to hang indefinitely. Similarly, without the size-based limit, a fast client that uploads hundreds of megabytes per second indefinitely will cause the server to download all of that data, hanging while it's doing so. Both protections are complementary, and since the implementation complexity of the time-based limit is low, I think it should stay.

If you think that NGINX (or Apache, etc). is a better place than WSGI-Kerberos to implement this type of logic, then this is argument for dropping WSGI-Kerberos's `_consume_request` altogether in favor of configuring NGINX to handle this, as I also argued in https://github.com/deshaw/wsgi-kerberos/pull/15#issuecomment-749762424 and implemented in #25. If WSGI-Kerberos immediately sent 401 responses without first waiting to drain the request socket for potentially naive clients, it would free an upstream proxy like NGINX to immediately see the 401 and then take responsibility for draining the request socket in a safe way before forwarding that 401 to the client. As long as WSGI-Kerberos is getting itself into the business of draining request sockets before even beginning to send its 401 responses, it defeats upstream proxies from handling this more safely, and requires WSGI-Kerberos to handle it safely for its users.

Reply:
To be clear, I still think #25 should be merged rather than this PR, and hope you’ll reconsider that after this discussion.

Reply:
Also, just using a production-quality WSGI server, such as Gunicorn or uWSGI, keeping things entirely within the pip/PyPI package ecosystem, is sufficient for getting safe `_consume_request` behavior. With no need to install and set up any additional, external reverse proxy like NGINX.

I’m guessing the `_consume_request` behavior was added to WSGI-Kerberos when testing it with a non-production-quality WSGI server, not realizing that switching to a production-quality WSGI server would have made the issue disappear, but also solved it in a safer and more robust way than it can be solved at the WSGI middleware layer (where the sockets have already been set up and are managed for us by the WSGI server).

Reply:
Had to resubmit #25 as #29 and hope that will be merged instead of this PR.

Reply:
The size limit is a bit different than the time limit. Enabling the size limit may be desirable specifically to encourage CLI clients that upload large files to enable pre-emptive authentication. The time limit is a generic DoS protection mechanism that could apply to any request - it's not really specific to wsgi_kerberos or Kerberos authenticated requests. That's why users concerned about this should configure the limit in the web server.

Adding another option (`read_timeout_on_auth_fail`) adds additional complexity and maintenance burden. Once a feature is out there, we need to be more careful about not changing or breaking it. In general, I want to avoid adding features unless there is a strong practical need for it.

(I would also be in favor of changing the `read_max_on_auth_fail` default to infinite, and configure the limit on the web-server if necessary. But its purpose is slightly different in this case, so I could see an argument for keeping it)

Reply:
In the latest revision, I've removed `read_timeout_on_auth_fail`, and retained your previously-suggested increase of the default `read_max_on_auth_fail` to 200 MB, which I think is a safer default than infinity.

Another look?